src/lua/version.lua

```diff
@@ -0,0 +1,259 @@
+-- using tables is not possible.
+--
+ffi.cdef([[
+    struct version {
+        /** Major number. */
+        uint32_t major;
+        /** Minor number. */
+        uint32_t minor;
+        /** Patch number. */
+        uint32_t patch;
+        /** Pre-release name id. */
+        uint8_t _prerelease_id;
+        /** Pre-release number: e.g. in alpha1 it's 1. */
+        uint16_t _prerelease_num;
+        /** Number of commits since release. */
+        uint32_t commit;
+    };
+]])
+
+--
+-- Pre-release names sorted in descending order.
+-- ID 0 is reserved for missing pre-release name.
+--
+local prerelease_names = {
+    'rc',
+    'beta',
+    'alpha',
+    'entrypoint',
+}
+
+local prerelease_name_to_id = {}
+for id, name in ipairs(prerelease_names) do
+    prerelease_name_to_id[name] = id
+end
+
+local function version_new(major, minor, patch, prerelease, commit)
+    commit = commit or 0
+    check_param(major, 'major', 'number', 2)
+    check_param(minor, 'minor', 'number', 2)
+    check_param(patch, 'patch', 'number', 2)
+    check_param(commit, 'commit', 'number', 2)
+
+    local prerelease_id
+    local prerelease_num = 0
+    if prerelease then
+        local prerelease_name
+        check_param(prerelease, 'prerelease', 'string', 2)
+        prerelease_name, prerelease_num = prerelease:match('^(%a+)(%d+)')
+        if not prerelease_name then
+            prerelease_name = prerelease:match('^(%a+)')
+        else
+            prerelease_num = tonumber(prerelease_num)
+        end
+        prerelease_id = prerelease_name_to_id[prerelease_name]
+        if not prerelease_id then
+            local msg = "Unknown release type '%s'"
+            box.error(box.error.ILLEGAL_PARAMS,
+                      string.format(msg, prerelease_name), 2)
+        end
+    else
+        -- Greatest priority, actual release.
+        prerelease_id = 0
+    end
+
+    assert(prerelease_id <= MAX_PRERELEASE_NAME_WEIGHT)
+    return ffi.new('struct version', {
+        major = major,
+        minor = minor,
+        patch = patch,
+        _prerelease_id = prerelease_id,
+        _prerelease_num = prerelease_num,
+        commit = commit,
+    })
+end
+
+local function version_fromstr(version_str)
+    check_param(version_str, 'version_str', 'string', 2)
+    --  x.x.x-name<num>-<num>-g<commit>
+    -- \____/\________/\_____/
+    --   P1      P2      P3
+    local major, minor, patch, prerelease, commit, pos
+
+    -- Part 1 - version ID triplet. Compulsory fields.
+    major, minor, patch = version_str:match('^(%d+)%.(%d+)%.(%d+)')
+    if not major or not minor or not patch then
+        local msg = "Error during parsing version string '%s'. " ..
+                    "At least X.Y.Z is required."
+        box.error(box.error.ILLEGAL_PARAMS, string.format(msg, version_str), 2)
+    end
+    major = tonumber(major)
+    minor = tonumber(minor)
+    patch = tonumber(patch)
+    -- Cut to 'type<num>-<num>-g<commit>'.
+    pos = version_str:find('-')
```

Comment:
This `pos` does not necessarily correspond to the `<major>.<minor>.<patch>` triplet end.

```lua
tarantool> version('3.3.0HAHAHAHAHA-entrypoint-202-gedf82e996d')
---
- 3.3.0-entrypoint-202
...
```

Reply:
I've decided to strictify the checks, since it's not difficult to do. Now arbitrary symbols between version parts in string are forbidden. Also, I've forbidden to add arbitrary info in the end of the version string, if it's not prefixed with `-` symbol: `3.3.0-rc1dev` is forbidden, `3.3.0-rc1-dev` is allowed.